pkg/route/secret/monitor_test.go

```diff
@@ -0,0 +1,354 @@
+package secret
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+func fakeMonitor(ctx context.Context, fakeKubeClient *fake.Clientset, key ObjectKey) *singleItemMonitor {
+	sharedInformer := fakeSecretInformer(ctx, fakeKubeClient, key.Namespace, key.Name)
+	return newSingleItemMonitor(key, sharedInformer)
+}
+
+// fakeSecretInformer will list/watch only one secret inside a namespace
+func fakeSecretInformer(ctx context.Context, fakeKubeClient *fake.Clientset, namespace, name string) cache.SharedInformer {
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", name).String()
+	klog.Info(fieldSelector)
+	return cache.NewSharedInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = fieldSelector
+			return fakeKubeClient.CoreV1().Secrets(namespace).List(ctx, options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = fieldSelector
+			return fakeKubeClient.CoreV1().Secrets(namespace).Watch(ctx, options)
+		},
+	},
+		&corev1.Secret{},
+		0,
+	)
+}
+
+func fakeSecret(namespace, name string) *corev1.Secret {
+	return &corev1.Secret{
+		Type: corev1.SecretTypeOpaque,
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"test": {1, 2, 3, 4},
+		},
+	}
+}
+
+func TestStartInformer(t *testing.T) {
+	scenarios := []struct {
+		name      string
+		isClosed  bool
+		expectErr bool
+	}{
+		{
+			name:      "pass closed channel into informer",
+			isClosed:  true,
+			expectErr: true,
+		},
+		{
+			name:      "pass unclosed channel into informer",
+			isClosed:  false,
+			expectErr: false,
+		},
+	}
+
+	for _, s := range scenarios {
+		t.Run(s.name, func(t *testing.T) {
+			fakeKubeClient := fake.NewSimpleClientset()
+			monitor := fakeMonitor(context.TODO(), fakeKubeClient, ObjectKey{})
+			if s.isClosed {
+				close(monitor.stopCh)
+			}
+			go monitor.StartInformer(context.TODO())
+			time.Sleep(10 * time.Millisecond)
+
+			select {
+			// this case will execute if stopCh is closed
+			case <-monitor.stopCh:
+				if !s.expectErr {
+					t.Error("informer is not running")
+				}
+			default:
+				t.Log("informer is running")
+			}
+		})
```

Comment:
I don't quite get what we are trying to test here. `StartInformer` calls the informer's `Run()` and `StopInformer` when the context is done. None of these is tested here.

Reply:
- `StartInformer` is a blocking call because it’s calling informer's `Run()`, and will only return when it’s channel is closed. So I think the only way to test whether the informer’s `Run()` is getting appropriately called or not, or whether the informer is running or not, is to play with it’s channel. 
- `case <-monitor.stopCh:` will be executed if the channel is closed and informer is not running. And if the channel is not closed, which means `StartInformer()` has successfully started the informer, `default`  will be executed in that case.
- We have a dedicated `TestStopWithContextCancel` test case to check the context cancellation scenario.